plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/DeployJob.java

```diff
@@ -216,9 +220,16 @@ private CloudSdk getCloudSdk(Path credentialFile,
         .exitListener(processExitListener)
         .appCommandMetricsEnvironment(CloudToolsInfo.METRICS_NAME)
         .appCommandMetricsEnvironmentVersion(CloudToolsInfo.getToolsVersion())
-        .appCommandOutputFormat("json")
-        .build();
-    return cloudSdk;
+        .appCommandOutputFormat("json");
+
+    IJavaProject javaProject = JavaCore.create(project);
```

Comment:
> and the DeployCommandHandler is already doing a lot of configuration.

I'm going to digress a bit, but I don't think this is really true, because what it does is merely pre-constructing a `CloudSdk`-specific class from the save project preferences. (BTW, I think is a not good design. `DeployJob` can completely hide `CloudSdk`-specific stuff.) Sure, it additionally passes a work directory, but Elliotte once said it shouldn't (he left a comment somewhere) and rather `DeployJob` should just create one inside. I agree on that. Then there's some message console arguments, but this is just a separate between the background Job and the UI command handler; we don't have other options but to pass the UI arguments. So, then it just boils down to passing the actual user deploy preferences.

We may find that the project JDK may not be sufficient for CloudSdk (e.g., may need Java 8).

Going back to the main discussion, that's a good point. Now I think we should have a separate `CloudSdk` instance, because the issue here is only for case of App Engine standard staging. And for that `CloudSdk`, I think it should always use the project-specific JDK, which shouldn't be configurable. What you pointed out is about running `CloudSdk` in general, and that could be made configurable (made sometime later, if a need arises). I'll do some refactoring.